File: leetcode/Arrays/leetcode-229/229.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        int target1, target2;
        int count1 = 0, count2 = 0;

        for (int num : nums) {
            if (num == target1) {
                count1++;
            } else if (num == target2) {
                count2++;
            } else if (count1 == 0) {
                target1 = num;
                count1 = 1;
            } else if (count2 == 0) {
                target2 = num;
                count2 = 1;
            } else {
                count1--;
                count2--;
            }
        }

        count1 = 0;
        count2 = 0;
        for (int num : nums) {
            if (num == target1) {
                count1++;
            } else if (num == target2) {
                count2++;
            }
        }

        List<Integer> res = new ArrayList<>();
        if (count1 > nums.length / 3) {
            res.add(target1);
        }
        if (count2 > nums.length / 3) {
            res.add(target2);
        }
        return res;
    }
}

File: leetcode/Arrays/leetcode-229/229.go
package arrays

func majorityElement(nums []int) []int {
	t1, c1, t2, c2 := 0, 0, 0, 0

	for _, num := range nums {
		if num == t1 {
			c1++
		} else if num == t2 {
			c2++
		} else if c1 == 0 {
			t1 = num
			c1 = 1
		} else if c2 == 0 {
			t2 = num
			c2 = 1
		} else {
			c1--
			c2--
		}
	}

	c1 = 0
	c2 = 0
	for _, num := range nums {
		if num == t1 {
			c1++
		} else if num == t2 {
			c2++
		}
	}

	var res []int
	if c1 > len(nums)/3 {
		res = append(res, t1)
	}
	if c2 > len(nums)/3 {
		res = append(res, t2)
	}
	return res
}
